Yield events for unknown tool calls

proever opened the pull request:
implements #1913 

Claude 4 Sonnet might have gone a bit overboard on the tests here. I am happy to trim them down a bit or move them into an existing file.

DouweM commented on pydantic_ai_slim/pydantic_ai/_agent_graph.py:
```diff
@@ -649,7 +649,11 @@ async def process_function_tools(  # noqa C901
+            yield _messages.FunctionToolCallEvent(call)
+
+            result = _unknown_tool(call.tool_name, call.tool_call_id, ctx)
+            yield _messages.FunctionToolResultEvent(result, tool_call_id=call.tool_call_id)
```
I think we'll want to do the same thing for the `ToolReturnPart` we output on line 645

DouweM commented:
@proever 

> Claude 4 Sonnet might have gone a bit overboard on the tests here. I am happy to trim them down a bit or move them into an existing file.

Yeah the tests are a bit much 😄 I'd suggest adding just a single test in `tests/test_streaming.py`, modeled after `tests/models/test_google.py::test_google_model_iter_stream`, that tests a combination of successful and failed tool calls. And we prefer using inline-snapshot to test the entire `all_events` list over specifically checking things like `all_events[1].result.tool_name == 'unknown_tool'`.

proever commented:
@DouweM would you mind taking another look? I've added events for output tool validation failures and simplified the tests (one for each event case and moved into `test_streaming`).

DouweM commented on tests/test_streaming.py:
```diff
@@ -921,3 +924,100 @@ def output_validator(data: OutputType | NotOutputType) -> OutputType | NotOutput
+
+    assert event_parts == snapshot(
+        [
+            IsInstance(FunctionToolCallEvent),
```
@proever Ah, interesting. What happens if you manually remove the `call_id` line from the snapshot and run the test again?

proever replied:
I did that in the second example, in both cases the error is the same (although I think the route to that error is different depending on if `call_id` is provided explicitly in the test or implicitly by `inline-snapshot`).

proever replied:
yeah I think that makes sense. Should be all done now.

Though I would argue for completeness' sake we should do something similar with `FunctionToolResultEvent`:

```py
@dataclass(repr=False)
class FunctionToolResultEvent:
    """An event indicating the result of a function tool call."""

    result: ToolReturnPart | RetryPromptPart
    """The result of the call to the function tool."""
    event_kind: Literal['function_tool_result'] = 'function_tool_result'
    """Event type identifier, used as a discriminator."""

    @property
    def tool_call_id(self) -> str:
        """An ID used to match the result to its original call."""
        return self.result.tool_call_id

    __repr__ = _utils.dataclasses_no_defaults_repr
```

Both `ToolReturnPart` and `RetryPromptPart` have a `tool_call_id` so it should work, but I think this one is a bit more work to update in the various places and tests it is used. Probably a separate PR. If you agree I am happy to do that some time, or we can leave things as is!

proever replied:
Agreed. I think a separate PR probably makes sense for that one though. 

Do you have any recommendations for the failing coverage check here? I'm not quite sure what the right approach is.

DouweM replied:
@proever I've added a `pragma: no branch` comment, let's see if that does it.

DouweM commented on tests/test_streaming.py:
```diff
@@ -921,3 +923,120 @@ def output_validator(data: OutputType | NotOutputType) -> OutputType | NotOutput
+
+    try:
+        async with agent.iter('test') as agent_run:
+            async for node in agent_run:
```
```suggestion
            async for node in agent_run:  # pragma: no branch
```
